src/identifier.rs

```diff
@@ -1,14 +1,42 @@
 //! Define the type of an identifier.
+use once_cell::sync::Lazy;
 use serde::{Deserialize, Serialize};
-use std::{fmt, hash::Hash};
+use std::fmt;
+use std::hash::Hash;
 
 use crate::position::TermPos;
 
-#[derive(Debug, Clone, Default, Deserialize, Serialize)]
+static INTERNER: Lazy<interner::Interner> = Lazy::new(interner::Interner::new);
+
+#[derive(Debug, Clone, Copy, Deserialize, Serialize)]
 #[serde(into = "String", from = "String")]
 pub struct Ident {
-    pub label: String,
-    pub pos: TermPos,
+    label: interner::Symbol,
+    pos: TermPos,
+    generated: bool,
+}
+
+impl Ident {
+    pub fn new_with_pos(label: impl AsRef<str>, pos: TermPos) -> Self {
+        let generated = label.as_ref().starts_with(GEN_PREFIX);
+        Self {
+            label: INTERNER.intern(label),
+            pos,
+            generated,
+        }
+    }
+
+    pub fn new(label: impl AsRef<str>) -> Self {
+        Self::new_with_pos(label, TermPos::None)
+    }
+
+    pub fn label(&self) -> &str {
+        INTERNER.lookup(self.label)
+    }
+
+    pub fn pos(&self) -> TermPos {
```

Comment:
Is there a good reason not to make `pos` public? For `symbol` it makes sense, but `pos` can be freely modified without consequences. I would tend to avoid adding a boilerplate `pos()` function, changing all the code that uses `.pos` somewhere, without a motivation to not use the language's accessor (random example: [codespan's diagnostic has all fields public](https://docs.rs/codespan-reporting/0.11.1/codespan_reporting/diagnostic/struct.Diagnostic.html))

Reply:
Okay, I'll answer to that with another question: does it make any sense to be able to modify `pos`? Using an access method makes it clear that we can only read the field. If it makes sense to be able to modify this field, then it should be reverted.
(I don't see the connection with `codespan`...)

Reply:
The connection with codespan is the following: they have the same choice to make for `Diagnostic`. They could go with method accessors instead, but they didn't. This is just one datapoint.

I tend to see types like Ident etc. as describing data. You should be able to manipulate data freely, as long as it doesn't have any bad consequences. If no methods and no other part of the code rely on the fact that `pos` is immutable, then I don't see a reason to impose this constraint, which moreover needs this additional accessor. For example, the `pos` of RichTerm is not a method, and we often modifes the position afterward, because we build them in several steps (granted, we usually use `with_pos`, but just because it's nicer to chain, and not always: the error handling code does modify the position of a rich term directly because it makes the overall code simpler, see e.g. https://github.com/tweag/nickel/blob/54bfa6eb3074305ab289cec98fe2d6bedd1fdaf2/src/error.rs#L878). On paper, we could also say that it doesn't make sense to modify the position of a `RichTerm`, but in practice it can come in handy.

It also makes things inconsistent with the rest of the codebase (for `RichTerm` you use `.pos`, for `Label` you use `.arg_pos`, but for `Ident` you use a function call `.pos()`).

Reply:
okay then, I'm changing this back